lib/giraph/subquery.rb

```diff
@@ -0,0 +1,72 @@
+module Giraph
+  # Defines a thin & sane API for the sub-query extraction
+  # from the context provided to resolvers via AST, graphql-ruby API
+  # and a touch of regex.
+  class Subquery
```

Comment:
Agreed. Also the reason why I put all this heavily-dependent logic into this class. At least it's isolated now and easier to follow/fix.

I always had the idea to propose an API change in mind, just wasn't sure exactly how I would end up tying this in and what I would use. Now that we have a working implementation, I think the next step is to think about how to make this more natural. What I need are:

- A way to override/wrap `resolve` more naturally rather then monkey-patching, and preferably per Type too,  vs. a global patch.

- A way to figure out the subquery string given a currently resolving field. This is not too bad right now, `context.ast_node.to_query_string` is okay, could be cleaner.

- A way to figure out variables used within an AST node, so that I can only declare those and send their values. Right now this is handled with regex, which is... let's say not reliable enough.

- Preferably a way to identify data_type (type + nullability) without having to resort to case'ing potentially wrapped types. Not too bad as is, could be better.

Essentially, anything that relies on AST directly is good enough as is and very unlikely to change. Other things can be done better.